README.txt
Kitserver 3.1.0 README FILE                                February 19, 2016
============================================================================


INTRODUCTION
------------

Kitserver 3 a "companion" program for Pro Evolution Soccer 3
and Winning Eleven 7 International. It enhances the "Strip Selection" menu 
of the game by allowing to select additional uniforms for Home and Away 
teams (both player and goalkeeper kits), and adds ball selection as well.
Configurable hot-keys are used to trigger selections.

This makes it easy to use "3rd" or "alternative" or "European" or 
"Champions League" (or whatever you want to call them) kits for specific 
matches, without having to edit the AFS-file and the executable.  Also, 
sometimes, teams use different shorts or socks with standard colored
shirts. This too is easily done with Kitserver: all you have to do is 
just add extra uniforms with new combinations. 
Another usage for Kitserver is the testing of new kits.

The uniforms are picked from a special database (KDB), which basically
contains the uniXXX*.bin files, and the description of number colors/collars
for each uniform. You can easily add new uniforms to the database by simply
copying files to KDB/players directory, and modifying attrib.cfg file to
set the correct colors and collars. Uniform files must be named like this:

uniXXX*.bin

which means that the name must start with "uni", followed by 3 digit number
of the team, and it must have extension ".bin". Anything else in the name
is optional. Examples:

uni020-allwhite.bin   -  correct name for a uniform file for Russia
uni6bla.bin           -  INCORRECT name: team name must be 3 digit number.
                         (correct name would be uni006bla.bin)
uni077pb.bin          -  correct name for a uniform file for Newcastle


!!!WARNING!!!:
Bad uniform files (corrupt or truncated) WILL CRASH the game, when you try
to select them with Kitserver. ALWAYS try a new kit in an exhibition game
first, before using it in your Master League campaign. (I did ran into a 
few bad uniform files that i downloaded for testing, so the bad kits do
exist unfortunately.)

The ball files are also stored in KDB. Unlike the kits, you MUST configure
every ball in the attrib.cfg file. See sample ball configuration in the
attrib.cfg. You can provide just the ball texture, or both the model file
and the texture file.


SYSTEM REQUIREMENTS
-------------------

Kitserver only works with PES3(PC) and WE7I(PC).
Supported versions are: 

- PES3 1.0 (original)
- PES3 1.10.2 (Konami patch)
- PES3 1.30.1 (Konami patch)
- WE7I US


INSTALLATION
------------

STEP 1. Copy the entire "kitserver" folder into your game folder. 
(NOTE: Don't rename "kitserver" folder to anything else, because that
will prevent the game from loading Kitserver)

STEP 2. Run setup.exe 
If STEP 1 was done correctly, you should see your "pes3.exe" (or "we7.exe")
in the dropdown list. You may see other executable files in that list, if
you have other executable files in the game folder (not counting setting.exe)
Select the executable, which you want to install the Kitserver for. 
If Kitserver hasn't been already installed for this executable, the "Install"
button should become enabled. Press "Install" button. The installation should
happen pretty quickly - in the matter of seconds. Once, it is complete, the
popup window will display "SUCCESS!" message, or report an error if one 
occured.

If you get an error message, verify that the game is not currently 
running. Kitserver cannot be installed, when the game is running.
Another typical cause for error, is a read-only executable file. In
Windows Explorer, right-click on the file, and select "Properties". Uncheck 
Read-Only Attribute checkbox, click "OK". Go back to setup window, and try
pressing "Install" button again.

STEP 3. Close setup.exe.
Installation is complete.


USAGE
-----

Just start the game, as you normally would.

When you get to the STRIP SELECTION screen, look for the blue star
near the middle of the screen. If you see it, it means that extra uniforms
are available for one of the teams (or both). To select these extra kits,
press one of the hot keys. If a team has many extra kits, every time you 
press the hot key - next kit is shown.
If you don't see the star indicator in the center of the screen, then it
means that no extra uniforms were found in KDB for either of your teams.

Default keys:
'1' - switch home player kit
'2' - switch away player kit
'3' - switch home goalkeeper kit
'4' - switch away goalkeeper kit
'B' - switch ball
'R' - switch to a random ball

You can re-assign these keys by using "kctrl.exe" program, which provides
a friendly GUI for Kitserver configuration. (In order to create this 
configuration file (kserv.cfg), you must run kctrl.exe, and press "Save" 
button.

ADVANCED USAGE:
kctrl.exe also allows to modify the path to your KDB folder. 
Normally, you don't need to worry about it. The only situation when you
may want to change it, is when you have 2 or more games that Kitserver
is compatible with (PES3, WE7I), and you want to share the KDB between
them. You still need to do the normal installation procedure for each
game, but after that is done, you can move KDB folder to some other 
place on your hard drive, and then make sure that all kitservers point
to it. This way, you can maintain just one database for all games.


KSERV.CFG OPTIONS
-----------------

Some additional features are available for configuration in kserv.cfg
file, which is located in the same folder where kserv.dll is.
Here is a short description of what each of those means:

game.speed

- defaults to 1.0
- allows to alter the game speed. Values less than 1.0 will make the
game slower, and values bigger than 1.0 will make the game faster.
Even small changes can make a noticeable difference, so if you want
to slow the game down just a little bit, use a value like 0.89, for example.

aspect.ratio

- sets the aspect ratio during the matches. Useful for widescreen modes,
because it allows to set correct proportions of the ball and players.
This is done by changing the horizontal viewing area of the camera.
For widescreen modes, you would want to set it to 1.6 or 1.77778, 
depending on exact resolution in use.

screen.width
screen.height

- allows to set custom resolution, which isn't offered by the game.
This can happen typically with newer monitors, especially widescreens.

camera.zoom

- defaults to 1800.0
- sets the value of camera zoom. Values less than 1800.0 will make the
field appear further away, values bigger than 1800.0 will bring the field
closer to the viewer.

camera.angle.multiplier

- defaults to 1.0
- allowed values are: 1,2,4,8,16,32
- Some cameras in the game provide the "angle", which determines how far
the camera can turn left and right. 0 (Default) means almost no turning at
all, and 9 - the biggest angle. With this option, you can increase that
angle. It works by basically multuplying the "angle" value in the game
by the value of this option. So, setting the multiplier to 2 or larger
values, will result in bigger turn angle for the camera.

stadium.render.clip

- defaults to 1
- allowed values are: 0,1
- Disables (0) or enables (1) the clipping of the stadium left and right
parts during the gameplay. Clipping is done by the game engine to optimize
the performance, but with modified camera zoom and angle, you can get 
Setting this option to 0 allows to turn off the clipping, so that the 
stadium left and right sides are always correctly rendered.

stadium.render.height

- defaults to 0.15
- Controls the height of rendering of the stadium. By default, during the
gameplay, the upper parts of the stadium are not rendered, because they
are typically not visible. By increasing this value, you make the game
engine render more of the stadium. A value of 1.0 or 2.0 should result in
the entire stadium being rendered (with the roof or upper structures).
Be aware though that near side will be rendered too, therefore 
obstructing the camera, on cameras such as "Wide" and "Long". Choose
the value carefully.


UNINSTALL
---------

STEP 1. Run setup.exe.
Choose the executable file that you wish to uninstall Kitserver from.
If Kitserver is currently installed for this executable, the "Remove"
button will be enabled. Press "Remove" button and you should see the
"SUCCESS!" pop-up window, if uninstall went ok. 

If error is reported, verify that the game is not currently running,
make sure the executable file is not READ-ONLY, and try again.

STEP 2. Close setup.exe
Uninstallation is complete.
(If you want to completely get rid of the Kitserver, you can now 
safely delete the "kitserver" folder).


SAMPLE KDB
----------

I included only few uniform files in the KDB - just to demonstrate how
the Kitserver works. It's easy to add many more. (see instructions above 
on how to do it). All you need to do is to find a good website with
uniforms and start building the KDB database. 

Some of the included player kits are:

uni067-3rd.bin        - Arsenal third kit (dark blue)
uni067pa-reds.bin     - Arsenal standard home kit, but with red socks
uni077-3rd.bin        - Newcastle grey (european) kit
uni120.bin            - Barcelona home kit
uni126_home_white.bin - Dinamo Kiev home kit
uni126_away_blue.bin  - Dinamo Kiev away kit

Also included are goalkeeper kits for Newcastle United and Bayern Munich.

There are 5 balls included in KDB. Read the "[Balls]" section in the
attrib.cfg file to see how they are configured.


CREDITS
-------

Kitserver 3 developed by Juce.
Beta-testing by biker_jim_uk.

Dinamo Kiev kits were created by ntalex and sch.
- http://www.schfootball.narod.ru/kits.html

All other kits in the KDB that is included with this program
were downloaded from the following sites:
- http://www.wekel.net (Kit author: kEL)
- http://www.pespc-files.com (Kit author: Spark)

Balls were created by ibo, Yogui, Sqwall.
(downloaded from http://www.pespc-files.com)


============================================================================


HISTORY.txt
KitServer CHANGE LOG                                       February 19, 2016
============================================================================


Versino 3.1.0
-------------
- new features to tweak camera settings:
    * camera.zoom
    * camera.angle.multiplier
    * stadium.render.clip
    * stadium.render.height


Version 3.0.1
-------------
- fixed problems introduced in 3.0.0
  for: PES3 1.10.2, 1.30.1 and WE7 International


Version 3.0.0
-------------
- new features:
    * game speed regulator
    * aspect ratio (good for playing on widescreen monitors)
    * custom screen resolution


Version 2.0.5
-------------
- new feature: random ball selection.


Version 2.0.4
-------------
- A major bug fixed, when goalkeeping kits were not loading properly:
(loading into wrong team, not loading at all, etc.) Also the number/name 
colors for GK kits are now correctly set.
- That PES3 1.10.2 bug is fixed again. For sure. :-)


Version 2.0.3
-------------
- added a couple extra sample kits to KDB (for Dinamo Kiev)
- Unfortunately, the PES 1.10.2 bug that was fixed in v2.0.1 
was accidently introduced back in this release.


Version 2.0.1
-------------
This is bug-fixing (a.k.a. "maintanence") release.
The following bugs are fixed:

- In MATCH MODE, "P1 quick start" and "P2 quick start" modes were not handled
correctly. This resulted in weird effect of kits from different teams being
selected for players and goalkeepers, if more then one match was played in
either of these two modes.

- For PES3 1.10.2, players kits were always loaded into "1st strip", instead
of being loaded into currently selected strip. (This bug does not appear
if PES3 1.0, 1.30.1 and WE7I executable is used.)


Version 2.0
-----------
This is almost a complete re-write of the KitServer. It now has a different
architecture, where you have to install it as a plug-in for the game, unlike 
in earlier (1.x) versions when you had to keep the kserv.exe running together
with the game so that it could hook on it.

The new architecture has several advantages:

1. Requires less resources, because no windows hooking is needed.
2. The game itself loads KitServer on startup, and unloads on exit, so no 
need for manual start of kserv.exe (in fact, there is no kserv.exe anymore)
3. Compatibility mode problems, which i believe were caused by the need
to do run-time hooking, should be resolved now.

The following features are also new in this release:

- Ball selection (You can select from any number of balls from KDB)
- Goalkeeper kit selection (KDB now contains goalkeeper kits as well)
- Text labels for all selectable items for easy identification. (For
example, some teams use the "European" kits, which look identical from the
front, but are different at the back. Text label identifies exactly which
kit file is currently in use.)




config.cpp
// config.cpp

#include <stdio.h>
#include <string.h>

#include "config.h"
#include "log.h"

/**
 * Returns true if successful.
 */
BOOL ReadConfig(KSERV_CONFIG* config, char* cfgFile)
{
	if (config == NULL) return false;

	bool hasKdbDir = false;
	FILE* cfg = fopen(cfgFile, "rt");
	if (cfg == NULL) return false;

	char str[BUFLEN];
	char name[BUFLEN];
	int value = 0;
	float dvalue = 0.0f;

	char *pName = NULL, *pValue = NULL, *comment = NULL;
	while (true)
	{
		ZeroMemory(str, BUFLEN);
		fgets(str, BUFLEN-1, cfg);
		if (feof(cfg)) break;

		// skip comments
		comment = strstr(str, "#");
		if (comment != NULL) comment[0] = '\0';

		// parse the line
		pName = pValue = NULL;
		ZeroMemory(name, BUFLEN); value = 0;
		char* eq = strstr(str, "=");
		if (eq == NULL || eq[1] == '\0') continue;

		eq[0] = '\0';
		pName = str; pValue = eq + 1;

		ZeroMemory(name, NULL); 
		sscanf(pName, "%s", name);

		if (lstrcmp(name, "debug")==0)
		{
			if (sscanf(pValue, "%d", &value)!=1) continue;
			LogWithNumber("ReadConfig: debug = (%d)\n", value);
			config->debug = value;
		}
		else if (lstrcmp(name, "kdb.dir")==0)
		{
			char* startQuote = strstr(pValue, "\"");
			if (startQuote == NULL) continue;
			char* endQuote = strstr(startQuote + 1, "\"");
			if (endQuote == NULL) continue;

			char buf[BUFLEN];
			ZeroMemory(buf, BUFLEN);
			memcpy(buf, startQuote + 1, endQuote - startQuote - 1);

			LogWithString("ReadConfig: kdbDir = \"%s\"\n", buf);
			ZeroMemory(config->kdbDir, BUFLEN);
			strncpy(config->kdbDir, buf, BUFLEN-1);

			hasKdbDir = true;
		}
		else if (lstrcmp(name, "vKey.homeKit")==0)
		{
			if (sscanf(pValue, "%x", &value)!=1) continue;
			LogWithNumber("ReadConfig: vKeyHomeKit = 0x%02x\n", value);
			config->vKeyHomeKit = value;
		}
		else if (lstrcmp(name, "vKey.awayKit")==0)
		{
			if (sscanf(pValue, "%x", &value)!=1) continue;
			LogWithNumber("ReadConfig: vKeyAwayKit = 0x%02x\n", value);
			config->vKeyAwayKit = value;
		}
		else if (lstrcmp(name, "vKey.gkHomeKit")==0)
		{
			if (sscanf(pValue, "%x", &value)!=1) continue;
			LogWithNumber("ReadConfig: vKeyGKHomeKit = 0x%02x\n", value);
			config->vKeyGKHomeKit = value;
		}
		else if (lstrcmp(name, "vKey.gkAwayKit")==0)
		{
			if (sscanf(pValue, "%x", &value)!=1) continue;
			LogWithNumber("ReadConfig: vKeyGKAwayKit = 0x%02x\n", value);
			config->vKeyGKAwayKit = value;
		}
		else if (lstrcmp(name, "vKey.ball")==0)
		{
			if (sscanf(pValue, "%x", &value)!=1) continue;
			LogWithNumber("ReadConfig: vKeyBall = 0x%02x\n", value);
			config->vKeyBall = value;
		}
		else if (lstrcmp(name, "vKey.randomBall")==0)
		{
			if (sscanf(pValue, "%x", &value)!=1) continue;
			LogWithNumber("ReadConfig: vKeyRandomBall = 0x%02x\n", value);
			config->vKeyRandomBall = value;
		}
		else if (lstrcmp(name, "aspect.ratio")==0)
		{
            float fvalue;
			if (sscanf(pValue, "%f", &fvalue)!=1) continue;
            if (fvalue >= 0.5f && fvalue <= 5.0f) {
                LogWithNumber("ReadConfig: aspectRatio = %03f\n", fvalue);
                config->aspectRatio = fvalue;
            }
		}
		else if (lstrcmp(name, "game.speed")==0)
		{
            float fvalue;
			if (sscanf(pValue, "%f", &fvalue)!=1) continue;
            if (fvalue >= 0.1f && fvalue <= 5.0f) {
                LogWithNumber("ReadConfig: gameSpeed = %03f\n", fvalue);
                config->gameSpeed = fvalue;
            }
		}
		else if (lstrcmp(name, "screen.width")==0)
		{
            DWORD dvalue;
			if (sscanf(pValue, "%u", &dvalue)!=1) continue;
            LogWithNumber("ReadConfig: screenWidth = %u\n", dvalue);
            config->screenWidth = dvalue;
		}
		else if (lstrcmp(name, "screen.height")==0)
		{
            DWORD dvalue;
			if (sscanf(pValue, "%u", &dvalue)!=1) continue;
            LogWithNumber("ReadConfig: screenHeight = %u\n", dvalue);
            config->screenHeight = dvalue;
		}

        // camera
		else if (lstrcmp(name, "camera.zoom")==0)
		{
            float fvalue;
			if (sscanf(pValue, "%f", &fvalue)!=1) continue;
            if (fvalue >= 10.0f && fvalue <= 10000.0f) {
                LogWithNumber("ReadConfig: cameraZoom = %03f\n", fvalue);
                config->cameraZoom = fvalue;
            }
		}
		else if (lstrcmp(name, "stadium.render.height")==0)
		{
            float fvalue;
			if (sscanf(pValue, "%f", &fvalue)!=1) continue;
            if (fvalue >= 0.1f && fvalue <= 5.0f) {
                LogWithNumber(
                    "ReadConfig: stadiumRenderHeight = %03f\n", fvalue);
                config->stadiumRenderHeight = fvalue;
            }
		}
		else if (lstrcmp(name, "stadium.render.clip")==0)
		{
            DWORD dvalue;
			if (sscanf(pValue, "%u", &dvalue)!=1) continue;
            if (dvalue == 0 || dvalue == 1) {
                LogWithNumber(
                    "ReadConfig: stadiumRenderClip = %u\n", dvalue);
                config->stadiumRenderClip = dvalue;
            }
		}
		else if (lstrcmp(name, "camera.angle.multiplier")==0)
		{
            DWORD dvalue;
			if (sscanf(pValue, "%u", &dvalue)!=1) continue;
            switch (dvalue) {
                case 1:
                case 2:
                case 4:
                case 8:
                case 16:
                case 32:
                    LogWithNumber(
                        "ReadConfig: cameraAngleMultiplier = %u\n",
                        dvalue);
                    config->cameraAngleMultiplier = dvalue;
                    break;
                default:
                    Log("ReadConfig: WARNING: camera.angle.multiplier"
                        " can only have these values: 1,2,4,8,16,32");
            }
		}

	}
	fclose(cfg);

	// if KDB directory is not specified, assume default
	if (!hasKdbDir)
	{
		strcpy(config->kdbDir, DEFAULT_KDB_DIR);
	}

	return true;
}

/**
 * Returns true if successful.
 */
BOOL WriteConfig(KSERV_CONFIG* config, char* cfgFile)
{
	char* buf = NULL;
	DWORD size = 0;

	// first read all lines
	HANDLE oldCfg = CreateFile(cfgFile, GENERIC_READ, 0, NULL,
			OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
	if (oldCfg != INVALID_HANDLE_VALUE)
	{
		size = GetFileSize(oldCfg, NULL);
		buf = (char*)HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, size);
		if (buf == NULL) return false;

		DWORD dwBytesRead = 0;
		ReadFile(oldCfg, buf, size, &dwBytesRead, NULL);
		if (dwBytesRead != size) 
		{
			HeapFree(GetProcessHeap(), 0, buf);
			return false;
		}
		CloseHandle(oldCfg);
	}

	// create new file
	FILE* cfg = fopen(cfgFile, "wt");

	// loop over every line from the old file, and overwrite it in the new file
	// if necessary. Otherwise - copy the old line.
	
	BOOL bWrittenKdbDir = false; BOOL bWrittenVKeyBall = false;
	BOOL bWrittenVKeyHomeKit = false; BOOL bWrittenVKeyAwayKit = false;
	BOOL bWrittenVKeyGKHomeKit = false; BOOL bWrittenVKeyGKAwayKit = false;
	BOOL bWrittenVKeyRandomBall = false;

	char* line = buf; BOOL done = false;
	char* comment = NULL;
	if (buf != NULL) while (!done && line < buf + size)
	{
		char* endline = strstr(line, "\r\n");
		if (endline != NULL) endline[0] = '\0'; else done = true;
		char* comment = strstr(line, "#");
		char* setting = NULL;

		if ((setting = strstr(line, "kdb.dir")) && 
			(comment == NULL || setting < comment))
		{
			fprintf(cfg, "kdb.dir = \"%s\"\n", config->kdbDir);
			bWrittenKdbDir = true;
		}
		else if ((setting = strstr(line, "vKey.homeKit")) && 
				 (comment == NULL || setting < comment))
		{
			fprintf(cfg, "vKey.homeKit = 0x%02x\n", config->vKeyHomeKit);
			bWrittenVKeyHomeKit = true;
		}
		else if ((setting = strstr(line, "vKey.awayKit")) && 
				 (comment == NULL || setting < comment))
		{
			fprintf(cfg, "vKey.awayKit = 0x%02x\n", config->vKeyAwayKit);
			bWrittenVKeyAwayKit = true;
		}
		else if ((setting = strstr(line, "vKey.gkHomeKit")) && 
				 (comment == NULL || setting < comment))
		{
			fprintf(cfg, "vKey.gkHomeKit = 0x%02x\n", config->vKeyGKHomeKit);
			bWrittenVKeyGKHomeKit = true;
		}
		else if ((setting = strstr(line, "vKey.gkAwayKit")) && 
				 (comment == NULL || setting < comment))
		{
			fprintf(cfg, "vKey.gkAwayKit = 0x%02x\n", config->vKeyGKAwayKit);
			bWrittenVKeyGKAwayKit = true;
		}
		else if ((setting = strstr(line, "vKey.ball")) && 
				 (comment == NULL || setting < comment))
		{
			fprintf(cfg, "vKey.ball = 0x%02x\n", config->vKeyBall);
			bWrittenVKeyBall = true;
		}
		else if ((setting = strstr(line, "vKey.randomBall")) && 
				 (comment == NULL || setting < comment))
		{
			fprintf(cfg, "vKey.randomBall = 0x%02x\n", config->vKeyRandomBall);
			bWrittenVKeyRandomBall = true;
		}
		else
		{
			// take the old line
			fprintf(cfg, "%s\n", line);
		}

		if (endline != NULL) { endline[0] = '\r'; line = endline + 2; }
	}

	// if something wasn't written, make sure we write it.
	if (!bWrittenKdbDir) 
		fprintf(cfg, "kdb.dir = \"%s\"\n", config->kdbDir);
	if (!bWrittenVKeyHomeKit)
		fprintf(cfg, "vKey.homeKit = 0x%02x\n", config->vKeyHomeKit);
	if (!bWrittenVKeyAwayKit)
		fprintf(cfg, "vKey.awayKit = 0x%02x\n", config->vKeyAwayKit);
	if (!bWrittenVKeyGKHomeKit)
		fprintf(cfg, "vKey.gkHomeKit = 0x%02x\n", config->vKeyGKHomeKit);
	if (!bWrittenVKeyGKAwayKit)
		fprintf(cfg, "vKey.gkAwayKit = 0x%02x\n", config->vKeyGKAwayKit);
	if (!bWrittenVKeyBall)
		fprintf(cfg, "vKey.ball = 0x%02x\n", config->vKeyBall);
	if (!bWrittenVKeyRandomBall)
		fprintf(cfg, "vKey.randomBall = 0x%02x\n", config->vKeyRandomBall);

	// release buffer
	HeapFree(GetProcessHeap(), 0, buf);

	fclose(cfg);

	return true;
}

